### PR TITLE
PARQUET-361: Add semver prerelease logic.

### PR DESCRIPTION
**Pull request by rdblue**
This also adds more versions where PARQUET-251 is fixed.

### CODE REVIEW

File: parquet-column/src/main/java/org/apache/parquet/CorruptStatistics.java [hunk 1]
```diff
@@ -41,6 +41,8 @@
   // the bug involved writing invalid binary statistics, so stats written prior to this
   // fix must be ignored / assumed invalid
   private static final SemanticVersion PARQUET_251_FIXED_VERSION = new SemanticVersion(1, 8, 0);
+  private static final SemanticVersion PARQUET_251_FIXED_START = new SemanticVersion(1, 5, 0, null, "cdh5.5.0", null);
```
Comment (isnotinvain) [inline parquet-column/src/main/java/org/apache/parquet/CorruptStatistics.java]: maybe rename these to `CDH_5_PARQUET_251_FIXED_START` and `CDH_5_PARQUET_251_FIXED_END`

File: parquet-common/src/main/java/org/apache/parquet/SemanticVersion.java [hunk 2]
```diff
@@ -141,7 +186,91 @@ public int hashCode() {
...
+    public int compareTo(NumberOrString that) {
+      // Numeric identifiers always have lower precedence than non-numeric identifiers.
+      int cmp = compareBooleans(that.isNumeric, this.isNumeric);
+      if (cmp != 0) {
+        return cmp;
+      }
+
+      if (isNumeric) {
+        // identifiers consisting of only digits are compared numerically
+        return this.number - that.number;
```
Comment (isnotinvain) [inline parquet-common/src/main/java/org/apache/parquet/SemanticVersion.java]: can you use Integer.compare() here, or just inline it as written here:
https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/lang/Integer.java#L1021

the x - y trick has issues w/ overflow

File: parquet-common/src/main/java/org/apache/parquet/SemanticVersion.java [hunk 3]
```diff
@@ -141,7 +186,91 @@ public int hashCode() {
...
+      // A larger set of pre-release fields has a higher precedence than a
+      // smaller set, if all of the preceding identifiers are equal
+      int size = Math.min(this.identifiers.size(), that.identifiers.size());
+      for (int i = 0; i < size; i += 1) {
+        int cmp = identifiers.get(i).compareTo(that.identifiers.get(i));
+        if (cmp != 0) {
+          return cmp;
+        }
+      }
+      return this.identifiers.size() - that.identifiers.size();
```
Comment (isnotinvain) [inline parquet-common/src/main/java/org/apache/parquet/SemanticVersion.java]: same here for compare()

**Comment by isnotinvain** [comment]
overall looks good, some minor comments

**Comment by isnotinvain** [comment]
+1

**Comment by rdblue** [comment]
Thanks for the review, Alex!

**Pull request closed without merging**